klippa/src/glyf_loca.rs

```diff
@@ -0,0 +1,304 @@
+    let mut max_offset: u32 = 0;
+
+    //TODO: support not_def_outline and drop_hints
+    for (_new_gid, old_gid) in &plan.new_to_old_gid_list {
+        match loca.get_glyf(*old_gid, &glyf) {
+            Ok(g) => {
+                let Some(glyph) = g else {
+                    subset_glyphs.push(Vec::new());
+                    continue;
+                };
+                let subset_glyph = subset_glyph(&glyph, plan);
+                let trimmed_len = subset_glyph.len();
+                max_offset += padded_size(trimmed_len) as u32;
+                subset_glyphs.push(subset_glyph);
+            }
+            _ => {
+                return Err(SubsetTableError(Glyf::TAG));
+            }
+        }
+    }
+
+    //TODO: support force_long_loca in the plan
+    let loca_format: u8 = if max_offset < 0x1FFFF { 0 } else { 1 };
+
+    let glyf_cap = estimate_subset_table_size(font, Glyf::TAG, plan);
+    let mut glyf_out = Vec::with_capacity(glyf_cap);
+
+    let loca_cap = estimate_subset_table_size(font, Loca::TAG, plan);
+    let mut loca_out: Vec<u8> = Vec::with_capacity(loca_cap);
+
+    if loca_format == 0 {
+        loca_out.extend_from_slice(&0_u16.to_be_bytes());
+        let mut offset: u16 = 0;
+        for g in &subset_glyphs {
+            let padded_len = padded_size(g.len());
+            offset += padded_len as u16;
+            let glyph_offset = offset >> 1;
+            loca_out.extend_from_slice(&glyph_offset.to_be_bytes());
+            glyf_out.extend_from_slice(g);
+            if padded_len > g.len() {
+                glyf_out.extend_from_slice(&[0]);
+            }
+        }
+    } else {
+        loca_out.extend_from_slice(&0_u32.to_be_bytes());
+        let mut offset: u32 = 0;
+        for g in &subset_glyphs {
+            offset += g.len() as u32;
+            loca_out.extend_from_slice(&offset.to_be_bytes());
+            glyf_out.extend_from_slice(g);
+        }
+    }
+
+    // As a special case when all glyph in the font are empty, add a zero byte to the table,
+    // so that OTS doesn’t reject it, and to make the table work on Windows as well.
+    // See https://github.com/khaledhosny/ots/issues/52
+    if glyf_out.is_empty() {
+        glyf_out.extend_from_slice(&[0]);
+    }
+
+    let Ok(head_out) = subset_head(&head, loca_format) else {
+        return Err(SubsetTableError(Head::TAG));
+    };
+
+    builder.add_raw(Glyf::TAG, glyf_out);
+    builder.add_raw(Loca::TAG, loca_out);
+    builder.add_raw(Head::TAG, head_out);
+    Ok(())
+}
+
+fn padded_size(len: usize) -> usize {
+    len + len % 2
+}
+
+fn subset_glyph(glyph: &Glyph, plan: &Plan) -> Vec<u8> {
+    //TODO: support set_overlaps_flag and drop_hints
+    match glyph {
+        Composite(comp_g) => subset_composite_glyph(comp_g, plan),
+        Simple(simple_g) => subset_simple_glyph(simple_g),
+    }
+}
+
+// TODO: drop_hints and set_overlaps_flag
+fn subset_simple_glyph(g: &SimpleGlyph) -> Vec<u8> {
```

Comment:
it feels to me like there could be some modest gains to be had here by not doing as much intermediate allocation... for instance maybe we could just be writing all the glyphs into a buffer directly? we'd need to make sure we were padding, and we might need to rewrite that buffer to remove the padding if we ended up choosing the long loca format, but it doesn't feel that tricky, although we'd need to profile to know if it was meaningful.

Reply:
These allocations are not hot in profiling and we're fine with current performance(so will explore ways to improve it in the future). Also when drop_hints option is enabled, it will become tricky if we write all glyphs together.